File: java/behavioral_pattern/strategy/ShoppingCart.java
package strategy;

import java.util.ArrayList;
import java.util.List;

public class ShoppingCart {
    private List<Item> items = new ArrayList<>();
    private PaymentStrategy paymentStrategy;

    public void addItem(Item item) {
        items.add(item);
    }

    public void setPaymentStrategy(PaymentStrategy paymentStrategy) {
        this.paymentStrategy = paymentStrategy;
    }

    public int calculateTotal() {
        int total = 0;
        for (Item item : items) {
            total += item.getPrice();
        }

        return total;
    }

    public void checkout() {
        int amount = calculateTotal();
        if (paymentStrategy == null) {
            System.out.println("Please select a payment method");
        }
        paymentStrategy.pay(amount);
    }
}

File: typescript/behavioral_pattern/src/strategy/ShoppingCart.ts
import Item from "./Item"
import PaymentStrategy from "./PaymentStrategy";

class ShoppingCart {
    private items: Item[] = [];
    private paymentStrategy: PaymentStrategy | undefined;

    public addItem(item: Item): void {
        this.items.push(item);
    }

    public setPaymentStrategy(paymentStrategy: PaymentStrategy): void {
        this.paymentStrategy = paymentStrategy;
    }

    public calculateTotal(): number {
        let total = 0;
        for (const item of this.items) {
            total += item.getPrice();
        }

        return total;
    }

    public checkOut(): void {
        const amount = this.calculateTotal();
        if (this.paymentStrategy === undefined) {
            console.log("Please select a payment method");
            return;
        }
        this.paymentStrategy.pay(amount);
    }
}

export default ShoppingCart;